src/Orleans/Core/Grain.cs

```diff
@@ -39,7 +36,7 @@ protected IGrainFactory GrainFactory
         /// </summary>
         protected IServiceProvider ServiceProvider 
         {
-            get { return Runtime?.ServiceProvider; }
+            get { return Data.ServiceProvider ?? Runtime?.ServiceProvider; }
```

Comment:
BTW, injecting the IServiceProvider should return this same instance (whether scopes are used or not), as if scopes are used, it would return the scoped SP, and if they aren't, it would return the main SP, that's why I thought the fallback seemed reasonable (even though it will never fallback until we provide the option to avoid scopes).